card/lib/card/env.rb

```diff
@@ -2,6 +2,57 @@ class Card
 
   # Card::Env can differ for each request; Card.config should not
   module Env
+
+    # session history helpers: we keep a history stack so that in the case of
+    # card removal we can crawl back up to the last un-removed location
+    module LocationHistory
+      include Card::Location
+
+      def location_history
+        #warn "sess #{session.class}, #{session.object_id}"
+        session[:history] ||= [card_path('')]
+        if session[:history]
+          session[:history].shift if session[:history].size > 5
+          session[:history]
+        end
+      end
+
+      def save_location card
+        return if Env.ajax? || !Env.html? || !card.known? || (card.codename == 'signin')
```

Comment:
not suggesting any major change right now, but the html? test makes me wonder if read events should be view specific.  (or if they should just be called from views?)

I do think the "signin" thing should be handled in the set somehow (at some point)